### PR TITLE
adding possibiliity to select hideable columns

### LINKED ISSUE #24: Add option to set hideable columns
Currently all of the columns are able to be hidden away. I wanted to be able to only allow hiding some of them.

Method like self.set_hideable_columns(['A', 'B']) could do that.

### PR DESCRIPTION
**Pull request by Najiva**

### CODE REVIEW

File: django_tables2_column_shifter/tables.py [hunk 1]
```diff
@@ -23,6 +23,8 @@ def __init__(self, *args, **kwargs):
             self.template_name = self.shifter_template
         else:
             self.template = self.shifter_template
+        """By default all of the columns are hideable"""
```
Comment (djk2) [inline django_tables2_column_shifter/tables.py:26]: For one-line comment use `#`
multi line strin `"""` should be usage for docstring (for dscription class or method)

File: django_tables2_column_shifter/tables.py [hunk 2]
```diff
@@ -35,6 +37,13 @@ def get_column_default_show(self):
         else:
             return self.column_default_show
 
+    def set_hideable_columns(self, hideable_columns):
+        """ Sets columns that are hideable """
+        self.hideable_columns = []
+        for c in self.columns.iterall():
```
Comment (djk2) [inline django_tables2_column_shifter/tables.py]: A good practice is not to use a one-letter variable. 
Code should be self explained.
I suggest to change `c` to `col` or `column`

File: django_tables2_column_shifter/tables.py [hunk 3]
```diff
@@ -35,6 +37,13 @@ def get_column_default_show(self):
...
+            if c.name in hideable_columns:
```
Comment (djk2) [inline django_tables2_column_shifter/tables.py]: More efficient way is:

```
self.hideable_columns =list(set([col.name for col in self.columns]) & set(hideable_columns))
```

File: django_tables2_column_shifter/templates/django_tables2_column_shifter/bootstrap2.html [hunk 4]
```diff
@@ -27,7 +27,7 @@
                     <span class="caret"></span>
                 </button>
                 <ul class="dropdown-menu" style="min-width:350px; padding:5px; cursor:pointer;">
-                    {% for column in table.columns %}
+                    {% for column in table.hideable_columns %}
```
Comment (djk2) [inline django_tables2_column_shifter/templates/django_tables2_column_shifter/bootstrap2.html]: The loop should always iterate over all columns.
Of all the columns, some are active and some are not.

Pseudo code:
```
{% for col in all_cols%}
         {% if column in table.hideable_columns%}
                ## Button is active (old functionality)
         {% else%}
                ## Position is inactive (gray and not clicable) 
```

For `in active` elements you can add extra css (like 'not_hideable`) class have easy access from JS or CSS to remove for example this elements if you need. 

**Note: JS script should ignore inactive elements**

**Review by djk2 — COMMENTED** [review]
Generally:
1. Please, can you add unit tests to cover new feature? 
2. Please, can you run all unit tests using `tox` to check back compatibility for all django and django-tables versions
3. Please update README.md
4. Please increase second number in version in `django_tables2_column_shifter/__init__.py` to `2.1.0`
5. Add info about new feature to CHANGELOG.rst

**Comment by Najiva** [comment]
I will review the comments and update the updated commit later this month.

**Comment by djk2** [comment]
Hi, thanks for your PR.
No problem. anytime.

**Comment by Najiva** [comment]
How to run the test? When running tox tests, the tests pass but there are errors in the summary. Do I have to have all of the interpreters it is being tested against?

image

File: django_tables2_column_shifter/tables.py (hunk 1 above)
Comment (Najiva) [reply django_tables2_column_shifter/tables.py:26]: Wanted to be consistent with comment above it:

"""Override init for set shifter template"""

**Comment by Najiva** [comment]
I added test and also a css file since it was not there yet.  Also updated the version and readme. Please check if its OK.

File: CHANGELOG.rst [hunk 5]
```diff
@@ -1,5 +1,11 @@
 CHANGELOG
 ===========
+v. 2.1.0
+--------
+
+    * Adding column_shifter default css file
+	* Adding possibility to set selected columns to be hideable explicitly.
```
Comment (djk2) [inline CHANGELOG.rst:7]: Incorrect indent, 
Line 7 should have the same indent as a line number 6.

File: django_tables2_column_shifter/tables.py (hunk 1 above)
Comment (djk2) [reply django_tables2_column_shifter/tables.py:26]: Yes, but first comment under the method declaration is a docstring, please draw attention on fact that in line 21. comment for developers was marked using `#`. A multi-line comment is reserved for docstrings. 
More: https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html

File: README.rst [hunk 6]
```diff
@@ -264,6 +269,14 @@ Customizing:
             self.column_default_show = ['column1', 'column2']
             return super(MyModelTable, self).get_column_default_show()
 
+4. By default all columns are hideable. If you would like to disallow some columns to be
+   be hidable, specify hideable columns explicitly in the Table class constructor using
```
Comment (djk2) [inline README.rst:273]: `to be be hidable` - word `be` is repeated.

File: django_tables2_column_shifter/tables.py [hunk 7]
```diff
@@ -23,6 +23,8 @@ def __init__(self, *args, **kwargs):
             self.template_name = self.shifter_template
         else:
             self.template = self.shifter_template
+        """By default all of the columns are hideable"""
+        self.hideable_columns = self.columns
```
Comment (djk2) [inline django_tables2_column_shifter/tables.py:27]: Please consider below: 

```
self.set_hideable_columns(
    kwargs.get('hideable_columns') or self.columns
)
```

In this way, developer will be can very easy override `hideable columns` creating an object of class. 
This approach is more flexible. 
From one side we still have default behavior if kwargs not contain
hideable_columns, from other side if argument was passed developer can override default logic. 


Note:
Current logic is not consistent, from users you demand use method `set_hideable_columns(...)`- this method has some logic and maybe in future this logic will be extend. But yourself you don't use your own method to set columns, and assign columns directly (you ommit your own logic)

File: README.rst [hunk 8]
```diff
@@ -123,6 +123,11 @@ How to Install:
             type="text/javascript"
             src="{% static "django_tables2_column_shifter/js/django_tables2_column_shifter.min.js" %}">
         </script>
+        <link
```
Comment (djk2) [inline README.rst:126]: I would rather to not add new static file to *column-shifter*. 
From my point of view, will be better if we add styles for not-hidable column as  a inline tag attribute like `style="...."` 

I propose: 

1. Extend table class with attribute `not_hidable_css`

```
class ColumnShiftTable(tables.Table):
    ....
    # Inline css style for not-hidable column (for button)
    not_hidable_column_css = "pointer-events:none; opacity:.6"
```

2. In template
```
    <li class="btn-shift-column"
         style="{% if not column in table.hideable_columns %}{{ table.not_hidable_column_css }}{% endif %}"
         data-td-class="{{ column.attrs.td.class }}"
         data-state="on"
         .......
    </li>
```

In this way, new static file won't be necessary and we keep reverse compatibility 
Of course, info about it should be added to readme.

**Review by djk2 — COMMENTED** [review]
I can see, one issue with this functionality. 
Please consider below use case: 

1. User (developer) used old version my library. 
2. User (before upgrade) in web browser hid few columns
3. Developer installed new version of `column-shifter` and set some columns as hideable. 
4. Now user can hide or show only this definied by developer columns
5. If user earlier hid some columns, and now those columns are inactive then user never again show this columns.

Conslusion: 
non hidable columns should be always visible.

**Comment by Najiva** [comment]
Hi. I believe it is like that. Non-hidable columns are visible. But you cannot hide them (because they are disabled). Maybe what you mean is that also the state should be in that case reset to Visible.

**Comment by djk2** [comment]
Hi `Najiva`, I still wait for your changes to PR. 
Can you replay for my remarks and push code forward?
Your last reply not explain my all doubts. 
Please, review all my comments and apply some of them. 
Thanks for attention.

**Comment by Najiva** [comment]
Hi,I am sorry I dont have any free time anymore.Najiva





On Monday, July 19, 2021, 8:17 AM, K2 ***@***.***> wrote:




Hi Najiva, I still wait for your changes to PR.
Can you replay for my remarks and push code forward?
Your last reply not explain my all doubts.
Please, review all my comments and apply some of them.
Thanks for attention.

**Comment by djk2** [comment]
Don't worry, 
I exactly know how is. 
I'm also always on the go and haven't any free time. 
Go back to me, when you will be have a moment. 
I will left your PR still open. 
To next time.

**Pull request closed without merging**